[18.0][MIG]account_invoice_triple_discount

matteotognini opened the pull request:
This is based on #1939 by @Borruso and adjusts only the group assignment in the test to avoid a dependency on the sale module.
Specifically, I removed the group assignment line, as it was the only reference to sale in the entire module.

chienandalu committed:
account_invoice_triple_discount: New module

pedrobaeza committed:
account_invoice_triple_discount: Some improvements

xavierjimenez committed:
account_invoice_triple_discount

rschnapka committed:
Translated using Weblate (German)

Currently translated at 100,0% (4 of 4 strings)

Translation: account-invoicing-11.0/account-invoicing-11.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-11-0/account-invoicing-11-0-account_invoice_triple_discount/de/

Nikul-Chaudhary committed:
[MIG] account_invoice_triple_discount v11 to v12

Commit:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (4 of 4 strings)

Translation: account-invoicing-12.0/account-invoicing-12.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-12-0/account-invoicing-12-0-account_invoice_triple_discount/pt_BR/

benwillig committed:
[MIG] account_invoice_triple_discount to v13

ferran-S73 committed:
[IMP] account_invoice_triple_discount: black, isort, prettier

dreispt committed:
[FIX] account_invoice_triple_discount: Developer Mode was required to use discounts

francesco-ooops committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (4 of 4 strings)

Translation: account-invoicing-14.0/account-invoicing-14.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-14-0/account-invoicing-14-0-account_invoice_triple_discount/it/

ramiadavid committed:
[MIG] account_invoice_triple_discount: Migration to 15.0

jabelchi committed:
Translated using Weblate (Catalan)

Currently translated at 100.0% (4 of 4 strings)

Translation: account-invoicing-15.0/account-invoicing-15.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-15-0/account-invoicing-15-0-account_invoice_triple_discount/ca/

ramiadavid committed:
[FIX] account_invoice_triple_discount: fix precision when compute tax with discount

(cherry picked from commit 5afdb213c8f87be17bb21401a45cbf05a3548aa9)

ramiadavid committed:
[MIG] account_invoice_triple_discount: Migration to 16.0

ramiadavid committed:
[FIX][16.0] account_invoice_triple_discount : Error when print reports

badbole committed:
Translated using Weblate (Croatian)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/hr/

adrianojprado committed:
Translated using Weblate (Portuguese (Brazil))

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/pt_BR/

Ivorra78 committed:
Translated using Weblate (Spanish)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/es/

SirAionTech committed:
[FIX] account_invoice_triple_discount: Discount rounding

grindtildeath committed:
[REF] account_invoice_triple_discount: Consolidate discount in std field

Until now, the triple discount feature did use the standard Odoo field
as the first discount field, adding only extra fields for the second
and the third discount. This implied we had to override any function
using discount to consider the other discounts properly.

By adding an extra field discount1 to store the first discount, it allows
to redefine the standard discount field to a computed field that will
consolidate the triple discount from the other fields, and avoid the
need to redefine anything relying on the discount field as it will
already consider the triple discounts.

lmignon committed:
[FIX] account_invoice_triple_discount: Fix tax recompute on closed fiscal year

When computing the taxes and the totals on account.move.line it's required to prevent calls to the write method on account.move or triggering recompute of unexpected fields while we assign a temporary value to the discount field to use the aggregated one bedore calling the base implementation. Otherwise the update of the discount field will invalidate totals and ... on related accoun.move and trigger calls to the write method of account.move. Any call to the write method will trigger a set of check to see if the balance is still right or check if you try to write on a record prior to the fiscal year close date or .... At the end, every time you try to access to the taxes on the records, some errors could be unexpeced raised due to the trick a temporarily changing the discount value of the field while computing others fields.

In the future we should uses dedicated fields for 3 discount fields and make the native discount field a computed one with readonly=False. This approach would simplify a lot the code. See https://github.com/OCA/account-invoicing/pull/1638

lmignon committed:
[IMP] account_invoice_triple_discount: Improves tests

Uses CommonCase to take benefit of proper env initialization and avoid useless tacking records creation

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/it/

legalsylvain committed:
[IMP] account_invoice_triple_discount: Simplify account.move.line algorithm

[FIX] account_invoice_triple_discount: populate discount1 field to avoid inconsistency if database contains account.move.line with not null discount before the installation of the module

[REF] account_invoice_triple_discount: Add GRAP as coauthor of the module

[FIX] account_invoice_triple_discount: Manage Correct creation of account move lines, where 'discount' field is set

jakobkrabbe committed:
Translated using Weblate (Swedish)

Currently translated at 100.0% (8 of 8 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/sv/

ramiadavid committed:
Add test

grindtildeath committed:
Write discount into discount 1 if not defined

In case this module is installed and sale module is also installed
but not sale_triple_discount, or in case sale_triple_discount is not
updated to its latest version, make sure that setting discount, without
having discount1 defined, will not mess with the triple discounts.

legalsylvain committed:
[FIX] account_invoice_fixed_discount, account_invoice_triple_discount:
- set modules as incompatible in the manifest. (For the time being, there is only a informative message in the README of the module account_invoice_fixed_discount)
- Run separated tests for both modules.

legalsylvain committed:
[FIX] account_invoice_triple_discount: remove digits on overloaded field 'discount' as the field is now a technical field that should not have limitation. For exemple in the default case where digits precision of Discount is 2, if discounts are 05%, 09% and 13% the main discount is 24.7885 % (and not 24.79)

legalsylvain committed:
[REF] account_invoice_triple_discount: Improve speed computation of the field 'discount' and reduce complexity

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (11 of 11 strings)

Translation: account-invoicing-16.0/account-invoicing-16.0-account_invoice_triple_discount
Translate-URL: https://translation.odoo-community.org/projects/account-invoicing-16-0/account-invoicing-16-0-account_invoice_triple_discount/it/

Borruso committed:
[IMP] account_invoice_triple_discount: pre-commit auto fixes

Borruso committed:
[MIG] account_invoice_triple_discount: Migration to 18.0

legalsylvain committed:
[REF] account_invoice_triple_discount : integrate triple.discount.mixin, previously in purchase_triple_discount module. (OCA/purchase-workflow)

legalsylvain committed:
[REF] account.move.line now inherit triple.discount.mixin model

legalsylvain committed:
[REF] account_invoice_triple_discount: single line writing

Commit:
[IMP]account_invoice_triple_discount:remove groups from tests

legalsylvain reviewed (commented):
Hi. Could you cherry pick this one ?

https://github.com/OCA/account-invoicing/pull/2006

Thanks !

matteotognini commented:
Hi @legalsylvain, I cherry picked that one .I'm still getting some errors in the tests saying there are incompatible modules. Do you happen to know if there's a way to fix it? Thanks a lot!

legalsylvain committed:
[FIX] account_invoice_fixed_discount, account_invoice_triple_discount:
- set modules as incompatible in the manifest. (For the time being, there is only a informative message in the README of the module account_invoice_fixed_discount)
- Run separated tests for both modules.

legalsylvain commented:
hi @matteonext. I think you missed a commit. I made a try, and with an extra commit (https://github.com/OCA/account-invoicing/pull/2023/commits/8a151a2084b4d6870e9722ef2662a1e5b666c7bb) CI is green.

legalsylvain reviewed (changes requested):
Hi. Could you cherry pick this one ?

https://github.com/OCA/account-invoicing/pull/2006

Thanks !

matteotognini commented:
Hi @legalsylvain thanks for the review and the advices. I cherry picked the commit you show me. CI is green now.

odooNextev reviewed (approved):
LGTM

legalsylvain reviewed (approved):
Thanks!

matteotognini commented:
Hi @pedrobaeza. Would it be possible for you to merge this? Thanks

pedrobaeza commented:
/ocabot migration account_invoice_triple_discount

Merging due to existing reviews

/ocabot merge nobump

pedrobaeza commented:
/ocabot migration account_invoice_triple_discount
